src/appleseed/renderer/modeling/bsdf/microfacetnormalmappinghelper.h

```diff
@@ -0,0 +1,550 @@
+template <typename BSDFImpl>
+MicrofacetNormalMappingHelper<BSDFImpl>::MicrofacetNormalMappingHelper(
+    const char*                     name,
+    const ParamArray&               params)
+  : BSDFImpl(name, params)
+{
+}
+
+template <typename BSDFImpl>
+void MicrofacetNormalMappingHelper<BSDFImpl>::sample(
+	SamplingContext&                    sampling_context,
+	const void*                         data,
+	const bool                          adjoint,
+	const bool                          cosine_mult,
+	const LocalGeometry&                local_geometry,
+	const foundation::Dual3f&           outgoing,
+	const int                           modes,
+	BSDFSample&                         sample) const
+{
+	// World space original shading normal.
+	foundation::Vector3f original_shading_normal(local_geometry.m_shading_point->get_original_shading_normal());
+	if (foundation::dot(outgoing.get_value(), original_shading_normal) <= 0.0f)
+		return;
+
+	// World space perturbed shading normal.
+	foundation::Vector3f perturbed_shading_normal(local_geometry.m_shading_point->get_shading_normal());
+
+	// Perturbed normal too similar to original. Tangent vector can not be constructed.
+	if (foundation::dot(original_shading_normal, perturbed_shading_normal) > 1.0f - 1e-5)
+	{
+		BSDFImpl::sample(
+			sampling_context,
+			data,
+			adjoint,
+			false,
+			local_geometry,
+			outgoing,
+			modes,
+			sample);
+		return;
+	}
+
+	// World space tangent.
+    foundation::Vector3f tangent = build_tangent(original_shading_normal, perturbed_shading_normal);
+
+    const float shadow_terminator_freq_mult = local_geometry.m_shading_point->get_object_instance().get_render_data().m_shadow_terminator_freq_mult;
+
+	float final_pdf = 0.0f;
+	if (lambda_p(perturbed_shading_normal, outgoing.get_value(), original_shading_normal) > sampling_context.next2<float>())
+	{
+		// Case: hit perturbed facet.
+		BSDFSample perturbed_facet_sample;
+		float perturbed_facet_pdf = 0.0f;
+		BSDFImpl::sample(
+			sampling_context,
+			data,
+			adjoint,
+			false,
+			local_geometry,
+			outgoing,
+			modes,
+			perturbed_facet_sample);
+
+		perturbed_facet_sample.m_value *=
+			shift_cos_in_fast(adot(perturbed_facet_sample.m_incoming.get_value(), perturbed_shading_normal), shadow_terminator_freq_mult); // cosine weight, non-adjoint <wo, wp>
+
+		perturbed_facet_pdf = perturbed_facet_sample.get_probability();
+
+		if (G1(perturbed_shading_normal, perturbed_facet_sample.m_incoming.get_value(), original_shading_normal, true) > sampling_context.next2<float>())
+		{
+			// Sampled incoming direction was reflected on tangent facet (see Fig 10, right).
+			// Calculate reflected of incoming direction to find out the direction which intersects perturbed facet
+			// and calculate shadowing.
+			foundation::Vector3f incoming_reflected =
+				foundation::normalize(perturbed_facet_sample.m_incoming.get_value() - 2.0f * foundation::dot(perturbed_facet_sample.m_incoming.get_value(), tangent) * tangent);
```

Comment:
Too wide.

Reply:
I have formatted `sample()` for width in general. This should be OK now.